File: BhaktiChatiOS/Sources/Core/Services/Analytics.swift
import Foundation
import PostHog

/// Thin wrapper over the PostHog SDK so the rest of the app never touches the raw client.
///
/// Call `configure()` once from `BhaktiChatAppRoot`'s startup. If no API key is configured
/// (see `Config/Secrets.xcconfig` → POSTHOG_API_KEY), `configure()` is a no-op and every
/// `capture`/`screen` call is silently ignored — the app runs fine with analytics off.
///
/// Mirrors Android's `util/Analytics.kt` — keep event names and property keys in sync.
enum Analytics {
    private static var enabled = false

    /// Initializes PostHog. Safe to call with a blank key (becomes a no-op).
    static func configure(apiKey: String, host: String, debug: Bool) {
        guard !apiKey.isEmpty else { return }
        let config = PostHogConfig(projectToken: apiKey, host: host)
        // We track screens manually per SwiftUI tab/route (this app has no
        // Activity-level screen boundary the SDK could infer automatically).
        config.captureScreenViews = false
        config.captureApplicationLifecycleEvents = true
        config.debug = debug
        PostHogSDK.shared.setup(config)
        enabled = true
    }

    /// Records a custom event with optional properties. No-op when analytics is disabled.
    static func capture(_ event: String, properties: [String: Any]? = nil) {
        guard enabled else { return }
        PostHogSDK.shared.capture(event, properties: properties)
    }

    /// Records a screen view. `name` should be a stable route/screen identifier.
    static func screen(_ name: String, properties: [String: Any]? = nil) {
        guard enabled else { return }
        PostHogSDK.shared.screen(name, properties: properties)
    }

    /// Associates subsequent events with a known user id (e.g. after sign-in).
    static func identify(distinctId: String, userProperties: [String: Any]? = nil) {
        guard enabled else { return }
        PostHogSDK.shared.identify(distinctId, userProperties: userProperties)
    }

    /// Clears the current identity (e.g. on sign-out) so events go to a fresh anonymous id.
    static func reset() {
        guard enabled else { return }
        PostHogSDK.shared.reset()
    }

    // MARK: - Semantic helpers (keep event names in one place)

    static func chatMessageSent(guideId: String?) {
        var properties: [String: Any] = [:]
        if let guideId { properties["guide_id"] = guideId }
        capture("chat_message_sent", properties: properties)
    }

    static func guideSelected(guideId: String) {
        capture("guide_selected", properties: ["guide_id": guideId])
    }

    static func aartiOpened(aartiId: String) {
        capture("aarti_opened", properties: ["aarti_id": aartiId])
    }

    static func choghadiyaOpened() {
        capture("choghadiya_opened")
    }

    static func divineImageGenerationStarted(mode: String) {
        capture("divine_image_generation_started", properties: ["mode": mode])
    }

    static func divineImageGenerated(mode: String) {
        capture("divine_image_generated", properties: ["mode": mode])
    }

    static func divineImageRegenerated(mode: String) {
        capture("divine_image_regenerated", properties: ["mode": mode])
    }

    static func divineImageSaved(mode: String) {
        capture("divine_image_saved", properties: ["mode": mode])
    }

    /// `target` is "instagram", "whatsapp", or "system".
    static func divineImageShared(target: String) {
        capture("divine_image_shared", properties: ["target": target])
    }

    // MARK: - Ads
    // `format` e.g. "banner"/"interstitial"; `placement` e.g. "aartis_list"/"choghadiya".

    static func adShown(format: String, placement: String) {
        capture("ad_shown", properties: ["format": format, "placement": placement])
    }

    static func adClicked(format: String, placement: String) {
        capture("ad_clicked", properties: ["format": format, "placement": placement])
    }
}

File: BhaktiChatAndroidNative/app/src/main/java/com/bhaktichat/app/util/Analytics.kt
package com.bhaktichat.app.util

import android.content.Context
import com.posthog.PostHog
import com.posthog.android.PostHogAndroid
import com.posthog.android.PostHogAndroidConfig

/**
 * Thin wrapper over the PostHog SDK so the rest of the app never touches the raw client.
 *
 * Call [init] once from [com.bhaktichat.app.BhaktiChatApplication.onCreate]. If no API key
 * is configured (see local.properties → posthog.apiKey), [init] is a no-op and every
 * [capture]/[screen] call is silently ignored — the app runs fine with analytics off, so
 * debug builds and CI don't need a key.
 *
 * Event/property naming: use snake_case event names and keep property keys stable — PostHog
 * treats every distinct name as a new event/property in the dashboard.
 */
object Analytics {
    @Volatile
    private var enabled = false

    /** Initializes PostHog. Safe to call with a blank key (becomes a no-op). */
    fun init(context: Context, apiKey: String, host: String, debug: Boolean) {
        if (apiKey.isBlank()) return
        val config = PostHogAndroidConfig(apiKey = apiKey, host = host).apply {
            // We track screens manually per Compose route (this app is single-Activity, so
            // the SDK's Activity-level screen capture can't distinguish routes).
            captureScreenViews = false
            captureDeepLinks = true
            captureApplicationLifecycleEvents = true
            this.debug = debug
        }
        PostHogAndroid.setup(context.applicationContext, config)
        enabled = true
    }

    /** Records a custom event with optional properties. No-op when analytics is disabled. */
    fun capture(event: String, properties: Map<String, Any>? = null) {
        if (!enabled) return
        PostHog.capture(event = event, properties = properties)
    }

    /** Records a screen view. [name] should be a stable route/screen identifier. */
    fun screen(name: String, properties: Map<String, Any>? = null) {
        if (!enabled) return
        PostHog.screen(screenTitle = name, properties = properties)
    }

    /** Ties subsequent events to a persistent person profile. Call once per successful auth. */
    fun identify(userId: String, email: String?, name: String?) {
        if (!enabled) return
        PostHog.identify(
            distinctId = userId,
            userProperties = buildMap {
                email?.let { put("email", it) }
                name?.let { put("name", it) }
            }
        )
    }

    // --- Semantic helpers (keep event names in one place) -----------------

    /** [method] is "google" or "access" (email/username + password). */
    fun authSucceeded(method: String) =
        capture("auth_succeeded", mapOf("method" to method))

    /** [reason] is a short machine code, e.g. "network_error", "no_credential", or the API's error code. */
    fun authFailed(method: String, reason: String) =
        capture("auth_failed", mapOf("method" to method, "reason" to reason))

    fun chatMessageSent(guideId: String?) =
        capture("chat_message_sent", buildMap { guideId?.let { put("guide_id", it) } })

    fun guideSelected(guideId: String) =
        capture("guide_selected", mapOf("guide_id" to guideId))

    fun aartiOpened(aartiId: String) =
        capture("aarti_opened", mapOf("aarti_id" to aartiId))

    fun choghadiyaOpened() =
        capture("choghadiya_opened")

    fun divineImageGenerationStarted(mode: String) =
        capture("divine_image_generation_started", mapOf("mode" to mode))

    fun divineImageGenerated(mode: String) =
        capture("divine_image_generated", mapOf("mode" to mode))

    fun divineImageRegenerated(mode: String) =
        capture("divine_image_regenerated", mapOf("mode" to mode))

    fun divineImageSaved(mode: String) =
        capture("divine_image_saved", mapOf("mode" to mode))

    /** [target] is "instagram", "whatsapp", or "system". */
    fun divineImageShared(target: String) =
        capture("divine_image_shared", mapOf("target" to target))

    fun reelOpened() = capture("reel_opened")

    fun reelViewed(reelId: String, feed: String) =
        capture("reel_viewed", mapOf("reel_id" to reelId, "feed" to feed))

    fun reelShared() = capture("reel_shared")

    fun reelSetAsStatus() = capture("reel_set_as_status")

    fun aartiPlayStarted(aartiId: String) =
        capture("aarti_play_started", mapOf("aarti_id" to aartiId))

    fun aartiPlayCompleted(aartiId: String, durationSeconds: Long) =
        capture("aarti_play_completed", mapOf("aarti_id" to aartiId, "duration_seconds" to durationSeconds))

    /** Fired when the user manually skips off a track (next/previous) before it finished. */
    fun aartiSkipped(aartiId: String, positionSeconds: Long) =
        capture("aarti_skipped", mapOf("aarti_id" to aartiId, "position_seconds" to positionSeconds))

    // --- Ads (stubs — call these from the AdMob callbacks when ads land) ---
    // [format] e.g. "banner"/"interstitial"/"rewarded"/"app_open";
    // [placement] e.g. "result_return"/"chat_exit"/"aartis_list".

    fun adShown(format: String, placement: String) =
        capture("ad_shown", mapOf("format" to format, "placement" to placement))

    fun adClicked(format: String, placement: String) =
        capture("ad_clicked", mapOf("format" to format, "placement" to placement))
}
